File: 알고리즘 정리/정렬/문제풀이/가장 큰 수.py
from functools import cmp_to_key

def compare(s1,s2):
    short_len=min(len(s1),len(s2))
    s1_temp=s1[:short_len]
    s2_temp=s2[:short_len]
    if s1_temp>s2_temp:
        return -1
    elif s1_temp<s2_temp:
        return 1

    s1_temp=s1
    s2_temp=s2
    while len(s1_temp)<4:
        s1_temp+=s1
    while len(s2_temp)<4:
        s2_temp+=s2
    if s1_temp > s2_temp:
        return -1
    elif s1_temp < s2_temp:
        return 1
    else:
        return 0

def solution(numbers):
    number=[]
    for num in numbers:
        number.append(str(num))
    number.sort(key=cmp_to_key(compare))
    answer = ''
    for num in number:
        answer+=num
    if answer[0]=='0':
        return '0'
    return answer

numbers=[6,10,2]
print(solution(numbers))

File: 알고리즘 정리/정렬/문제풀이/가장 큰 수.cpp
#include<iostream>
///////////////////////////////////////////////////
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
string temp[100001];
bool compare(string a, string b)
{
	int len = a.size() < b.size() ? a.size() : b.size();
	if (a.substr(0, len).compare(b.substr(0, len)))
		return a > b;

	string s1 = a, s2 = b;
	while (s1.size() < 4)
		s1 += s1;
	while (s2.size() < 4)
		s2 += s2;
	return s1.substr(0, 4) > s2.substr(0, 4);
}

string solution(vector<int> numbers) {
	string answer = "";
	bool isEnd = true;
	for (int i = 0; i < numbers.size(); i++)
	{
		if (numbers[i] != 0)
			isEnd = false;
		temp[i] = to_string(numbers[i]);
	}
	if (isEnd)
		return "0";

	sort(temp, temp + numbers.size(), compare);
	for (int i = 0; i < numbers.size(); i++)
		answer += temp[i];
	return answer;
}

///////////////////////////////////////////////////

int main()
{
	cin.tie(0); cout.tie(0);
	cin.sync_with_stdio(0);

	vector<int> numbers = { 6,10,2 };
	cout<<solution(numbers);
	return 0;
}